worlds/ahit/Rules.py

```diff
@@ -0,0 +1,988 @@
+    "Battle of the Birds - Finale B": ["Battle of the Birds - Finale A"],
+
+    "Subcon Forest - Finale": ["Subcon Forest - Act 1", "Subcon Forest - Act 2",
+                               "Subcon Forest - Act 3", "Subcon Forest - Act 4",
+                               "Subcon Forest - Act 5"],
+
+    "The Arctic Cruise - Act 2":  ["The Arctic Cruise - Act 1"],
+    "The Arctic Cruise - Finale": ["The Arctic Cruise - Act 2"],
+}
+
+
+def can_use_hat(state: CollectionState, world: "HatInTimeWorld", hat: HatType) -> bool:
+    if world.options.HatItems.value > 0:
+        return state.has(hat_type_to_item[hat], world.player)
+
+    return state.count("Yarn", world.player) >= get_hat_cost(world, hat)
+
+
+def get_hat_cost(world: "HatInTimeWorld", hat: HatType) -> int:
+    cost = 0
+    costs = world.get_hat_yarn_costs()
+    for h in world.get_hat_craft_order():
+        cost += costs[h]
+        if h == hat:
+            break
+
+    return cost
+
+
+def can_sdj(state: CollectionState, world: "HatInTimeWorld"):
+    return can_use_hat(state, world, HatType.SPRINT)
+
+
+def painting_logic(world: "HatInTimeWorld") -> bool:
+    return world.options.ShuffleSubconPaintings.value > 0
+
+
+# -1 = Normal, 0 = Moderate, 1 = Hard, 2 = Expert
+def get_difficulty(world: "HatInTimeWorld") -> Difficulty:
+    return Difficulty(world.options.LogicDifficulty.value)
+
+
+def has_paintings(state: CollectionState, world: "HatInTimeWorld", count: int, allow_skip: bool = True) -> bool:
+    if not painting_logic(world):
+        return True
+
+    if world.options.NoPaintingSkips.value == 0 and allow_skip:
+        # In Moderate there is a very easy trick to skip all the walls, except for the one guarding the boss arena
+        if get_difficulty(world) >= Difficulty.MODERATE:
+            return True
+
+    return state.count("Progressive Painting Unlock", world.player) >= count
+
+
+def zipline_logic(world: "HatInTimeWorld") -> bool:
+    return world.options.ShuffleAlpineZiplines.value > 0
+
+
+def can_use_hookshot(state: CollectionState, world: "HatInTimeWorld"):
+    return state.has("Hookshot Badge", world.player)
+
+
+def can_hit(state: CollectionState, world: "HatInTimeWorld", umbrella_only: bool = False):
+    if world.options.UmbrellaLogic.value == 0:
+        return True
+
+    return state.has("Umbrella", world.player) or not umbrella_only and can_use_hat(state, world, HatType.BREWING)
+
+
+def can_surf(state: CollectionState, world: "HatInTimeWorld"):
+    return state.has("No Bonk Badge", world.player)
+
+
+def has_relic_combo(state: CollectionState, world: "HatInTimeWorld", relic: str) -> bool:
+    return state.has_group(relic, world.player, len(world.item_name_groups[relic]))
+
+
+def get_relic_count(state: CollectionState, world: "HatInTimeWorld", relic: str) -> int:
+    return state.count_group(relic, world.player)
+
+
+# Only use for rifts
+def can_clear_act(state: CollectionState, world: "HatInTimeWorld", act_entrance: str) -> bool:
```

Comment:
This is used for determining if the player can clear an act given an entrance name. It grabs the connected region from the entrance and checks the rule on the act completion location. Since free roam levels have no Time Piece associated with them, they're always considered cleared. This function is only used to determine if the player can clear an act that's required to unlock a Time Rift. I've renamed the function to `can_clear_required_act` and changed the comment above the function a bit if that helps.